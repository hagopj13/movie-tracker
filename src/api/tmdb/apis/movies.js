// @flow
import axios from '../axiosInstance';

export type GetDiscoverMoviesParams = {
  page?: number,
  sort_by: string,
  'vote_count.gte'?: number,
  with_genres?: string,
  'primary_release_date.gte'?: string,
  'primary_release_date.lte'?: string,
};
export const getDiscoverMovies = (params: GetDiscoverMoviesParams) =>
  axios.get('/discover/movie', { params });

export type GetUpcomingMoviesParams = {
  page?: number,
};
export const getUpcomingMovies = (params: GetUpcomingMoviesParams) =>
  axios.get('/movie/upcoming', { params });

export type SearchMoviesParams = {
  page?: number,
  query: string,
};
export const searchMovies = (params: SearchMoviesParams) => axios.get('/search/movie', { params });

export const getMovieDetails = (id: string, sessionId?: string) =>
  axios.get(`/movie/${id}`, {
    params: {
      session_id: sessionId,
      append_to_response: 'credits,keywords,reviews,recommendations,account_states',
    },
  });

export const getMovieUserState = (id: string, sessionId: string) =>
  axios.get(`/movie/${id}/account_states?session_id=${sessionId}`);

export const setIsMovieFavorite = (
  id: string,
  isFavorite: boolean,
  sessionId: string,
  accountId: string,
) =>
  axios.post(`/account/${accountId}/favorite?session_id=${sessionId}`, {
    media_type: 'movie',
    media_id: id,
    favorite: isFavorite,
  });

export const setIsMovieInWatchlist = (
  id: string,
  isInWatchlist: boolean,
  sessionId: string,
  accountId: string,
) =>
  axios.post(`/account/${accountId}/watchlist?session_id=${sessionId}`, {
    media_type: 'movie',
    media_id: id,
    watchlist: isInWatchlist,
  });

export const rateMovie = (id: string, rating: number, sessionId: string) =>
  axios.post(`/movie/${id}/rating?session_id=${sessionId}`, { value: rating });

export const deleteMovieRating = (id: string, sessionId: string) =>
  axios.delete(`/movie/${id}/rating?session_id=${sessionId}`);
