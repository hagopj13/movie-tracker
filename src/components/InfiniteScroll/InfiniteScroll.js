// @flow
import React, { useEffect, useCallback } from 'react';
import type { Node } from 'react';

type Props = {
  children: Node,
  activationDistance?: number,
  onLoadMore: () => void,
};

const InfiniteScroll = (props: Props) => {
  const { children, activationDistance, onLoadMore } = props;

  const handleScroll = useCallback(() => {
    const documentHeight = document.body?.offsetHeight ?? 0;
    const scrollHeight = window.innerHeight + window.scrollY;
    if (scrollHeight + activationDistance >= documentHeight) {
      onLoadMore();
    }
  }, [activationDistance, onLoadMore]);

  useEffect(() => {
    window.addEventListener('scroll', handleScroll);
    return () => window.removeEventListener('scroll', handleScroll);
  }, [handleScroll]);

  return <>{children}</>;
};

InfiniteScroll.defaultProps = {
  activationDistance: 20,
};

export default InfiniteScroll;
