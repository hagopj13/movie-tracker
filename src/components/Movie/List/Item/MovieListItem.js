// @flow
import React from 'react';
import { useHistory } from 'react-router-dom';
import moment from 'moment';
import { makeStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardActionArea from '@material-ui/core/CardActionArea';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import Rating from '@material-ui/lab/Rating';

import MoviePoster from 'components/Movie/Poster/MoviePoster';
import type { Movie } from 'types';

type Props = {
  movie: Movie,
};

const useStyles = makeStyles((theme) => ({
  root: {
    width: 200,
    minWidth: 200,
    cursor: 'pointer',
  },
  title: {
    fontSize: 16,
  },
  rating: {
    display: 'flex',
    marginTop: theme.spacing(0.75),
  },
  ratingText: {
    marginLeft: theme.spacing(0.5),
  },
}));

const MovieListItem = (props: Props) => {
  const { movie } = props;

  const classes = useStyles();
  const history = useHistory();

  const formattedDate = movie.releaseDate
    ? moment(movie.releaseDate).format('MMM D, YYYY')
    : 'Unknown release date';

  const handleItemClick = () => {
    history.push(`/movies/${movie.id}`);
  };

  const renderRating = () => {
    return movie.voteCount ? (
      <>
        <Rating value={movie.voteAverage} precision={0.5} readOnly />
        <Typography className={classes.ratingText}>{movie.voteAverage}</Typography>
      </>
    ) : (
      <Typography>Not rated yet</Typography>
    );
  };

  return (
    <Card className={classes.root} elevation={3} onClick={handleItemClick}>
      <CardActionArea>
        <MoviePoster posterPath={movie.posterPath} height={300} width={200} />
        <CardContent>
          <Typography className={classes.title} gutterBottom variant="h6">
            {movie.title}
          </Typography>
          <Typography>{formattedDate}</Typography>
          <div className={classes.rating}>{renderRating()}</div>
        </CardContent>
      </CardActionArea>
    </Card>
  );
};

export default MovieListItem;
