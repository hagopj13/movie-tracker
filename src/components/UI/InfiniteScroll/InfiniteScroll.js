// @flow
import React, { useEffect, useCallback } from 'react';
import type { Node } from 'react';

type Props = {
  children: Node,
  onLoadMore: () => void,
};

const InfiniteScroll = (props: Props) => {
  const { children, onLoadMore } = props;

  const activationDistance = 20;
  const handleScroll = useCallback(() => {
    const documentHeight = document.body?.offsetHeight;
    const scrollHeight = window.innerHeight + window.scrollY;
    if (scrollHeight + activationDistance >= documentHeight) {
      onLoadMore();
    }
  }, [onLoadMore]);

  useEffect(() => {
    window.addEventListener('scroll', handleScroll);
    return () => window.removeEventListener('scroll', handleScroll);
  }, [handleScroll]);

  return <>{children}</>;
};

export default InfiniteScroll;
