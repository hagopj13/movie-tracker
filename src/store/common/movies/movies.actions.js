// @flow
import typeGenerator from 'store/common/utils/typeGenerator';
import actionCreatorGenerator from 'store/common/utils/actionCreatorGenerator';
import type { Movie, Results } from 'types';

import MoviesActionTypes from './movies.types';

const fetchMovies = (namespace) => () => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MOVIES),
});

const fetchMoviesStart = (namespace) => () => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MOVIES_START),
});

const fetchMoviesSuccess = (namespace) => (results: Results<Movie>) => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MOVIES_SUCCESS),
  payload: { results },
});

const fetchMoviesFailure = (namespace) => (error: string) => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MOVIES_FAILURE),
  payload: { error },
});

const fetchMoreMovies = (namespace) => () => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MORE_MOVIES),
});

const fetchMoreMoviesStart = (namespace) => () => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MORE_MOVIES_START),
});

const fetchMoreMoviesSuccess = (namespace) => (results: Results<Movie>) => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MORE_MOVIES_SUCCESS),
  payload: { results },
});

const fetchMoreMoviesFailure = (namespace) => (error: string) => ({
  type: typeGenerator(namespace, MoviesActionTypes.FETCH_MORE_MOVIES_FAILURE),
  payload: { error },
});

type Action = {
  type: string,
  payload?: any,
};

type MoviesActionCreators = {
  fetchMovies: () => Action,
  fetchMoviesStart: () => Action,
  fetchMoviesSuccess: (results: Results<Movie>) => Action,
  fetchMoviesFailure: (error: string) => Action,
  fetchMoreMovies: () => Action,
  fetchMoreMoviesStart: () => Action,
  fetchMoreMoviesSuccess: (results: Results<Movie>) => Action,
  fetchMoreMoviesFailure: (error: string) => Action,
};

export default (namespace: string): MoviesActionCreators =>
  actionCreatorGenerator<MoviesActionCreators>(namespace, {
    fetchMovies,
    fetchMoviesStart,
    fetchMoviesSuccess,
    fetchMoviesFailure,
    fetchMoreMovies,
    fetchMoreMoviesStart,
    fetchMoreMoviesSuccess,
    fetchMoreMoviesFailure,
  });
