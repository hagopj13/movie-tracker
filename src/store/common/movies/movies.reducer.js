// @flow
import reducerGenerator from 'store/common/utils/reducerGenerator';
import type { Movie } from 'types';

import MoviesActionTypes from './movies.types';

type State = {
  list: Movie[],
  pagination: {
    page: number,
    totalPages: number,
    totalResults: number,
  } | null,
};

export const defaultInitialState: State = {
  list: [],
  pagination: null,
};

const fetchMoviesStart = () => defaultInitialState;

const fetchMoviesSuccess = (state: State, action) => ({
  ...state,
  list: action.payload.results.list,
  pagination: action.payload.results.pagination,
});

const fetchMoreMoviesSuccess = (state: State, action) => ({
  ...state,
  list: state.list.concat(action.payload.results.list),
  pagination: {
    ...state.pagination,
    page: action.payload.results.pagination.page,
  },
});

const moviesActionHandler = {
  [MoviesActionTypes.FETCH_MOVIES_START]: fetchMoviesStart,
  [MoviesActionTypes.FETCH_MOVIES_SUCCESS]: fetchMoviesSuccess,
  [MoviesActionTypes.FETCH_MORE_MOVIES_SUCCESS]: fetchMoreMoviesSuccess,
};

export default (namespace: string, initialState?: State) =>
  reducerGenerator(namespace, moviesActionHandler, initialState ?? defaultInitialState);
