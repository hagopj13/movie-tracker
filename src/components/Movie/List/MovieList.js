// @flow
import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';

import InfiniteScroll from 'components/InfiniteScroll/InfiniteScroll';
import type { Movie } from 'types';

import MovieListItem from './Item/MovieListItem';

type Props = {
  movies: Movie[],
  // eslint-disable-next-line react/no-unused-prop-types
  center?: boolean,
  onLoadMore?: () => void,
};

const useStyles = makeStyles((theme) => ({
  root: {
    display: 'flex',
    flexWrap: 'wrap',
    justifyContent: (props: Props) => (props.center ? 'center' : 'flex-start'),
    margin: theme.spacing(-4, 0, 0, -4),
    '& > *': {
      margin: theme.spacing(4, 0, 0, 4),
    },
    [theme.breakpoints.down('xs')]: {
      justifyContent: 'center',
    },
  },
  noResultsText: {
    fontWeight: theme.typography.fontWeightRegular,
    textAlign: (props: Props) => (props.center ? 'center' : 'left'),
  },
}));

const MovieList = (props: Props) => {
  const { movies, onLoadMore } = props;

  const classes = useStyles(props);

  if (movies.length === 0) {
    return (
      <Typography className={classes.noResultsText} variant="h6">
        No movies found
      </Typography>
    );
  }

  const renderList = () => (
    <div className={classes.root}>
      {movies.map((movie) => (
        <MovieListItem key={movie.id} movie={movie} />
      ))}
    </div>
  );

  if (!onLoadMore) {
    return renderList();
  }

  return <InfiniteScroll onLoadMore={onLoadMore}>{renderList()}</InfiniteScroll>;
};

MovieList.defaultProps = {
  center: true,
  onLoadMore: undefined,
};

export default MovieList;
