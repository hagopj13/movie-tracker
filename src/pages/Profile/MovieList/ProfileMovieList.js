// @flow
import React, { useEffect, useState } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Container from '@material-ui/core/Container';

import MovieList from 'components/Movie/List/MovieList';
import Spinner from 'components/Spinner/Spinner';
import type { Movie } from 'types';

type Props = {
  movies: Movie[],
  isLoading: boolean,
  isLoadingMore: boolean,
  onFetchMovies: () => void,
  onFetchMoreMovies: () => void,
};

const useStyles = makeStyles((theme) => ({
  root: {
    display: 'flex',
    justifyContent: 'center',
  },
  container: {
    padding: theme.spacing(5, 5, 2),
  },
}));

export const ProfileMovieList = (props: Props) => {
  const { movies, isLoading, isLoadingMore, onFetchMovies, onFetchMoreMovies } = props;

  const [showView, setShowView] = useState(false);
  const classes = useStyles();

  // Show the view after some timeout, because of lagging tab switch when view is shown instantly
  useEffect(() => {
    const timeout = setTimeout(() => {
      onFetchMovies();
      setShowView(true);
    }, 200);
    return () => clearTimeout(timeout);
  }, [onFetchMovies]);

  if (!showView) {
    return null;
  }

  if (isLoading) {
    return <Spinner />;
  }

  return (
    <div className={classes.root}>
      <Container className={classes.container}>
        <MovieList movies={movies} onLoadMore={onFetchMoreMovies} />
        {isLoadingMore && <Spinner />}
      </Container>
    </div>
  );
};

export default ProfileMovieList;
