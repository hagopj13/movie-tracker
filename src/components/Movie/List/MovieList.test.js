import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';
import Typography from '@material-ui/core/Typography';

import movies from 'store/fixtures/movies';
import InfiniteScroll from 'components/InfiniteScroll/InfiniteScroll';

import MovieList from './MovieList';
import MovieListItem from './Item/MovieListItem';

describe('MovieList component', () => {
  let shallow;
  let wrapper;

  const mockLoadMore = jest.fn();

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    jest.clearAllMocks();
    const mockProps = {
      movies: movies.list,
      onLoadMore: mockLoadMore,
    };
    wrapper = shallow(<MovieList {...mockProps} />);
  });

  it('should render the MovieList component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render MovieListItem for each movie', () => {
    expect(wrapper.find(MovieListItem)).toHaveLength(movies.list.length);
    expect(wrapper.find(MovieListItem).at(0).prop('movie')).toBe(movies.list[0]);
  });

  it('should render the movies inside InfiniteScroll if onLoadMore is defined', () => {
    expect(wrapper.exists(InfiniteScroll)).toBe(true);
  });

  it('should call onLoadMore when onLoadMore is called on InfiniteScroll', () => {
    wrapper.find(InfiniteScroll).prop('onLoadMore')();
    expect(mockLoadMore).toHaveBeenCalledTimes(1);
  });

  it('should render a message when no movies are found', () => {
    const mockProps = {
      movies: [],
      onLoadMore: mockLoadMore,
    };
    const newWrapper = shallow(<MovieList {...mockProps} />);
    expect(newWrapper.exists(Typography)).toBe(true);
  });

  it('should not render InfiniteScroll but still render the movies when onLoadMore is not defined', () => {
    const mockProps = {
      movies: movies.list,
      onLoadMore: null,
    };
    const newWrapper = shallow(<MovieList {...mockProps} />);
    expect(newWrapper.exists(InfiniteScroll)).toBe(false);
    expect(wrapper.find(MovieListItem)).toHaveLength(movies.list.length);
  });
});
