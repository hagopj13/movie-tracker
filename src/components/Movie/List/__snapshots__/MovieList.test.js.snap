// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList component should render the MovieList component correctly 1`] = `
<InfiniteScroll
  activationDistance={20}
  onLoadMore={[MockFunction]}
>
  <div
    className="makeStyles-root-1 makeStyles-root-3"
  >
    <MovieListItem
      key="419704"
      movie={
        Object {
          "backdropPath": "/5BwqwxMEjeFtdknRV792Svo0K1v.jpg",
          "id": 419704,
          "overview": "The near future, a time when both hope and hardships drive humanity to look to the stars and beyond. While a mysterious phenomenon menaces to destroy life on planet Earth, astronaut Roy McBride undertakes a mission across the immensity of space and its many perils to uncover the truth about a lost expedition that decades before boldly faced emptiness and silence in search of the unknown.",
          "posterPath": "/xBHvZcjRiWyobQ9kxBhO6B2dtRI.jpg",
          "releaseDate": "2019-09-17",
          "title": "Ad Astra",
          "voteAverage": 3,
          "voteCount": 3260,
        }
      }
    />
    <MovieListItem
      key="545609"
      movie={
        Object {
          "backdropPath": "/1R6cvRtZgsYCkh8UFuWFN33xBP4.jpg",
          "id": 545609,
          "overview": "Tyler Rake, a fearless mercenary who offers his services on the black market, embarks on a dangerous mission when he is hired to rescue the kidnapped son of a Mumbai crime lordâ€¦",
          "posterPath": "/wlfDxbGEsW58vGhFljKkcR5IxDj.jpg",
          "releaseDate": "2020-04-24",
          "title": "Extraction",
          "voteAverage": 3.8,
          "voteCount": 1404,
        }
      }
    />
    <MovieListItem
      key="668203"
      movie={
        Object {
          "backdropPath": "/wuJpiGNGv7wLg52ONizLrOugYp9.jpg",
          "id": 668203,
          "overview": "[RUMORED]  The sixth and final installment of the Scary Movie franchise that ignores the fifth film. A parody of Scream 4 and various other horror movies, sequels and reboots that were released between the late 2010's and 2020. Takes place 15 years after the fourth film.",
          "posterPath": "/mWOkjqqzWiaFC65i2nFuMzM9jne.jpg",
          "releaseDate": "2021-04-02",
          "title": "Scary Movie 6",
          "voteAverage": 0,
          "voteCount": 0,
        }
      }
    />
  </div>
</InfiniteScroll>
`;
